magma/primitives/memory.py

```diff
@@ -0,0 +1,105 @@
+        )
+        self.verilog_name = "coreir_mem"
+        self.coreir_name = "mem"
+        self.coreir_lib = "coreir"
+        self.coreir_genargs = {"width": width, "depth": depth,
+                               "has_init": init is not None}
+        self.coreir_configargs = {}
+        if init is not None:
+            raise NotImplementedError("CoreIR's memory primitive does not "
+                                      "compile init arg to verilog yet")
+            # self.coreir_configargs["init"] = <convert to json arg>
+
+        def _simulate(self, value_store, state_store):
+            cur_clk = value_store.get_value(self.clk)
+
+            if not state_store:
+                state_store['mem'] = [
+                    BitVector[width](0) for _ in range(depth)
+                ]
+                state_store['prev_clk'] = cur_clk
+
+            prev_clk = state_store['prev_clk']
+            clk_edge = cur_clk and not prev_clk
+            rdata = value_store.get_value(self.rdata)
+
+            if clk_edge:
+                index = int(
+                    BitVector[addr_width](value_store.get_value(self.raddr))
+                )
+                rdata = state_store['mem'][index].as_bool_list()
+
+            if clk_edge:
+                if value_store.get_value(self.wen):
+                    index = int(
+                        BitVector[addr_width](value_store.get_value(self.waddr))
+                    )
+                    state_store['mem'][index] = BitVector[width](
+                        value_store.get_value(self.wdata)
+                    )
+
+            state_store['prev_clk'] = cur_clk
+            value_store.set_value(self.rdata, rdata)
+
+        self.simulate = _simulate
+
+
+class Memory(Generator2):
+    def __init__(self, height, T: Kind,
+                 read_latency: int = 0, read_only: bool = False):
+        if read_latency < 0:
+            raise ValueError("read_latency cannot be negative")
+        addr_width = clog2(height)
+        data_width = T.flat_length()
+        # For ROM, we use coreir's generic mem and tie off the write signals
+        self.io = IO(
+            RADDR=In(Bits[addr_width]),
+            RDATA=Out(T),
+        ) + ClockIO()
+        if not read_only:
+            self.io += IO(
+                WADDR=In(Bits[addr_width]),
+                WDATA=In(T),
+                WE=In(Enable)
+            )
+            waddr = self.io.WADDR
+            wdata = as_bits(self.io.WDATA)
+            wen = self.io.WE
+        else:
+            waddr = Bits[addr_width](0)
+            wdata = Bits[data_width](0)
+            wen = Bit(0)
+        coreir_mem = CoreIRMemory(height, data_width)()
+        coreir_mem.waddr @= waddr
+        coreir_mem.wdata @= wdata
+        coreir_mem.wen @= wen
+        coreir_mem.raddr @= self.io.RADDR
+        rdata = from_bits(T, coreir_mem.rdata)
+        for _ in range(read_latency):
+            rdata = Register(T)()(rdata)
```

Comment:
Is there a way that the read_latency can be passed to CoreIR? I don't remember if this approach, a 0-latency (aka asynchronous) memory followed by a register in a separate module, will synthesize to the right result in Vivado, a BRAM with the right settings, rather than a lower clock rate solution, such as a BRAM followed by a register.

Reply:
Should be hard to implement, but not sure if @rdaly525 meant to keep it out of the memory primitive's parameter set?

Reply:
An alternative, but slightly more complex approach would be to run a pass that collects this pattern (register chain after memory read port) and packs it into a special module that generates the desired verilog (registers inside the memory module) so it's more of a technology specific code generation option

Reply:
I think a good pattern, which has already been followed for register primitives is to have keep the coreir memory "primitive" latency agnostic and have a "mantle" library generator which has a latency parameter. This can then be lowered into [coreir_mem |-> coreir_reg] or [fpga_bram(latency = 1)].

Reply:
The ideal approach in my mind is like @rsetaluri said, to keep the latency parameter in the mantle library generator. If the generator itself is staged, then we can lower it to appropriate verilog or fpga bram just using specialization on that generator. If things are not staged, then a pass to look for memories followed by registers would be needed.

Reply:
That being said I do understand the utility of a set of coreir primitives that are parameterized by latency. If we do that approach, we should do a separate set of generators  from the base set (like coreir_w_latency.add)

Reply:
Where will the fpga_bram primitive suggested by @rsetaluri come from? CoreIR?

Reply:
@David-Durst I meant it more to generically say we should raise to mantle library, then lower to desired backends. I'm not clear specifically how to lower mem w/ latency to fpga.

Is there a specific verilog structure which you know would produce the desired fpga bram in the presence of latency? Or do we have to instance the fpga primitive directly?

Reply:
I think we could do either. Ideally we would just instantiate the appropriate BRAM primitive for the particular tech lib (which FPGA) but we could also go the route of just producing verilog that the FPGA tools can interpret as a BRAM.

Reply:
In my experience, BRAMs are instanced using the right verilog in Vivado, not instantiation with a TCL call to create a BRAM IP. This is the approach used by chisel at https://www.chisel-lang.org/chisel3/memories.html#syncreadmem-sequentialsynchronous-read-sequentialsynchronous-write.

The verilog looks like https://github.com/David-Durst/embeddedHaskellAetherling/blob/79c5f6101f1c91ad2d9a180bd8a8ed6e39124b28/test/verilog_examples/aetherling_copies/big_real_32_conv2d/big_real_32_conv2d_2073600.v#L3974

Reply:
@David-Durst would you just be looking for a synchronous read (registered read port) module rather than a general latency parameter? That would be a simpler requirement to meet versus the more general latency parameter solution.

Reply:
@leonardt Yep, that's exactly what I want.

Reply:
@rdaly525 @rsetaluri Looking at Chisel, it looks like they have two variants of their memory primitive, comb/async read and synch read.  Would you be opposed to adding a "synchronous_read" boolean flag to the memory primitive? This would significantly reduce the required code changes to achieve David's requirements (we can just augment the magma primitive to use the synch read parameter for latency 1, and chain registers for greater than 1.  and the verilog code generation can handle the synchronous read logic).

This is one possible middle ground between keeping all latency related logic in the mantle memory generator and adding latency as a parameter to the memory primitive.

Chisel's motivation for the two variants are:
> These SyncReadMems will likely be synthesized to technology SRAMs (as opposed to register banks).

> Writes to Mems are combinational/asynchronous-read, sequential/synchronous-write. These Mems will likely be synthesized to register banks, since most SRAMs in modern technologies (FPGA, ASIC) tend to no longer support combinational (asynchronous) reads.

So one argument for having sync mem be a primitive (or parameter of the primitive) is that it would be a common technology target.